From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megi@xff.cz>
Date: Wed, 29 Jul 2026 02:23:45 +0200
Subject: media: ov5640: Fix runtime PM reference leak on volatile control read

Each of the three cases returned the error directly, leaking the
runtime PM reference taken at the top of the function and leaving the
sensor powered up until the next resume/suspend cycle. Remember the
error instead and always drop the reference.

Signed-off-by: Ondrej Jirman <megi@xff.cz>
---
 drivers/media/i2c/ov5640.c | 18 ++++++----
 1 file changed, 11 insertions(+), 7 deletions(-)

diff --git a/drivers/media/i2c/ov5640.c b/drivers/media/i2c/ov5640.c
index 111111111111..222222222222 100644
--- a/drivers/media/i2c/ov5640.c
+++ b/drivers/media/i2c/ov5640.c
@@ -3604,6 +3604,7 @@ static int ov5640_g_volatile_ctrl(struct v4l2_ctrl *ctrl)
 {
 	struct v4l2_subdev *sd = ctrl_to_sd(ctrl);
 	struct ov5640_dev *sensor = to_ov5640_dev(sd);
+	int ret = 0;
 	int val;
 
 	/* v4l2_ctrl_lock() locks our own mutex */
@@ -3615,26 +3616,29 @@ static int ov5640_g_volatile_ctrl(struct v4l2_ctrl *ctrl)
 	case V4L2_CID_AUTOGAIN:
 		val = ov5640_get_gain(sensor);
 		if (val < 0)
-			return val;
-		sensor->ctrls.gain->val = val;
+			ret = val;
+		else
+			sensor->ctrls.gain->val = val;
 		break;
 	case V4L2_CID_EXPOSURE_AUTO:
 		val = ov5640_get_exposure(sensor);
 		if (val < 0)
-			return val;
-		sensor->ctrls.exposure->val = val;
+			ret = val;
+		else
+			sensor->ctrls.exposure->val = val;
 		break;
 	case V4L2_CID_AUTO_FOCUS_STATUS:
 		val = ov5640_get_af_status(sensor);
 		if (val < 0)
-			return val;
-		sensor->ctrls.af_status->val = val;
+			ret = val;
+		else
+			sensor->ctrls.af_status->val = val;
 		break;
 	}
 
 	pm_runtime_put_autosuspend(&sensor->i2c_client->dev);
 
-	return 0;
+	return ret;
 }
 
 static int ov5640_s_ctrl(struct v4l2_ctrl *ctrl)
-- 
Armbian

